File: internal/datetime/parser.go
package datetime

import (
	"fmt"
	"strings"
	"time"
)

// Parser handles parsing of various date/time input formats
type Parser struct {
	config *DateTimeConfig
}

// NewParser creates a new Parser with the given configuration
func NewParser(config *DateTimeConfig) *Parser {
	if config == nil {
		config = DefaultConfig()
	}
	return &Parser{config: config}
}

// ParseDateTime attempts to parse a date/time string using multiple formats
func (p *Parser) ParseDateTime(input string) (time.Time, error) {
	if input == "" {
		return time.Time{}, NewDateTimeError(
			ErrInvalidFormat,
			"empty date/time input",
			input,
			nil,
		)
	}

	// Clean up the input
	input = strings.TrimSpace(input)

	// Try parsing with all supported formats
	parsed, err := p.ParseWithFormats(input, CommonInputFormats)
	if err != nil {
		return time.Time{}, NewDateTimeError(
			ErrInvalidFormat,
			fmt.Sprintf("unable to parse date/time: expected formats like '2006-01-02T15:04:05Z' or '01/02/2006 3:04 PM', got '%s'", input),
			input,
			err,
		)
	}

	// Apply default timezone if the parsed time doesn't have timezone info
	// Check if input explicitly specified timezone
	// For "-", only consider it a timezone if it appears after "T" (timezone offset, not date separator)
	hasTimezone := strings.Contains(input, "Z") || strings.Contains(input, "+") ||
		(strings.Contains(input, "T") && strings.LastIndex(input, "-") > strings.LastIndex(input, "T")) ||
		strings.HasSuffix(strings.ToUpper(input), "UTC")

	if !hasTimezone {
		// Apply default timezone - interpret the parsed time as being in the default timezone
		loc, err := time.LoadLocation(p.config.DefaultTimezone)
		if err != nil {
			return time.Time{}, NewDateTimeError(
				ErrInvalidTimezone,
				fmt.Sprintf("invalid default timezone: %s", p.config.DefaultTimezone),
				input,
				err,
			)
		}
		parsed = time.Date(
			parsed.Year(), parsed.Month(), parsed.Day(),
			parsed.Hour(), parsed.Minute(), parsed.Second(), parsed.Nanosecond(),
			loc,
		)
	}

	return parsed, nil
}

// ParseDate parses a date-only string and returns a time.Time at midnight in the default timezone
func (p *Parser) ParseDate(input string) (time.Time, error) {
	if input == "" {
		return time.Time{}, NewDateTimeError(
			ErrInvalidFormat,
			"empty date input",
			input,
			nil,
		)
	}

	input = strings.TrimSpace(input)

	// Date-only formats
	dateFormats := []string{
		"2006-01-02",
		"01/02/2006",
		"1/2/2006",
		"January 2, 2006",
		"Jan 2, 2006",
		"2 January 2006",
		"2 Jan 2006",
	}

	parsed, err := p.ParseWithFormats(input, dateFormats)
	if err != nil {
		return time.Time{}, NewDateTimeError(
			ErrInvalidFormat,
			fmt.Sprintf("unable to parse date: expected formats like '2006-01-02' or '01/02/2006', got '%s'", input),
			input,
			err,
		)
	}

	// Set to midnight in the default timezone
	loc, err := time.LoadLocation(p.config.DefaultTimezone)
	if err != nil {
		return time.Time{}, NewDateTimeError(
			ErrInvalidTimezone,
			fmt.Sprintf("invalid default timezone: %s", p.config.DefaultTimezone),
			input,
			err,
		)
	}

	return time.Date(
		parsed.Year(), parsed.Month(), parsed.Day(),
		0, 0, 0, 0, loc,
	), nil
}

// ParseTime parses a time-only string and combines it with the given date
func (p *Parser) ParseTime(input string, date time.Time) (time.Time, error) {
	if input == "" {
		return time.Time{}, NewDateTimeError(
			ErrInvalidFormat,
			"empty time input",
			input,
			nil,
		)
	}

	input = strings.TrimSpace(input)

	// Time-only formats
	timeFormats := []string{
		"15:04:05",
		"15:04",
		"3:04:05 PM",
		"3:04 PM",
		"3:04:05PM",
		"3:04PM",
	}

	// Parse the time component
	parsed, err := p.ParseWithFormats(input, timeFormats)
	if err != nil {
		return time.Time{}, NewDateTimeError(
			ErrInvalidFormat,
			fmt.Sprintf("unable to parse time: expected formats like '15:04' or '3:04 PM', got '%s'", input),
			input,
			err,
		)
	}

	// Combine with the provided date in the date's timezone
	return time.Date(
		date.Year(), date.Month(), date.Day(),
		parsed.Hour(), parsed.Minute(), parsed.Second(), parsed.Nanosecond(),
		date.Location(),
	), nil
}

// ParseWithFormats tries to parse the input with multiple format strings
func (p *Parser) ParseWithFormats(input string, formats []string) (time.Time, error) {
	var lastErr error

	for _, format := range formats {
		parsed, err := time.Parse(format, input)
		if err == nil {
			return parsed, nil
		}
		lastErr = err
	}

	return time.Time{}, lastErr
}

// ParseDateTimeWithTimezone parses a date/time string and applies the specified timezone
func (p *Parser) ParseDateTimeWithTimezone(input, timezone string) (time.Time, error) {
	// First parse the date/time
	parsed, err := p.ParseDateTime(input)
	if err != nil {
		return time.Time{}, err
	}

	// If timezone is specified, convert to that timezone
	if timezone != "" {
		loc, err := time.LoadLocation(timezone)
		if err != nil {
			return time.Time{}, NewDateTimeError(
				ErrInvalidTimezone,
				fmt.Sprintf("invalid timezone: %s", timezone),
				input,
				err,
			)
		}

		// Check if input explicitly specified timezone
		// For "-", only consider it a timezone if it appears after "T" (timezone offset, not date separator)
		hasTimezone := strings.Contains(input, "Z") || strings.Contains(input, "+") ||
			(strings.Contains(input, "T") && strings.LastIndex(input, "-") > strings.LastIndex(input, "T")) ||
			strings.HasSuffix(strings.ToUpper(input), "UTC")

		if hasTimezone {
			// If the parsed time already has timezone info, convert it to the specified timezone
			parsed = parsed.In(loc)
		} else {
			// Interpret the parsed time as being in the specified timezone
			parsed = time.Date(
				parsed.Year(), parsed.Month(), parsed.Day(),
				parsed.Hour(), parsed.Minute(), parsed.Second(), parsed.Nanosecond(),
				loc,
			)
		}
	}

	return parsed, nil
}

File: lambda/saml_auth/datetime/parser.js
/**
 * Parser handles parsing of various date/time input formats
 * Equivalent to internal/datetime/parser.go
 */

import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc.js';
import timezone from 'dayjs/plugin/timezone.js';
import customParseFormat from 'dayjs/plugin/customParseFormat.js';
import { DateTimeConfig, defaultConfig, DateTimeError, newDateTimeError, ERROR_TYPES, COMMON_INPUT_FORMATS } from './types.js';

// Configure Day.js plugins
dayjs.extend(utc);
dayjs.extend(timezone);
dayjs.extend(customParseFormat);

export class Parser {
    constructor(config) {
        this.config = config || defaultConfig();
    }
    
    /**
     * Attempts to parse a date/time string using multiple formats
     * @param {string} input - The date/time string to parse
     * @returns {Date} - Parsed Date object
     * @throws {DateTimeError} - If parsing fails
     */
    parseDateTime(input) {
        if (!input || typeof input !== 'string') {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                'empty date/time input',
                input,
                null
            );
        }
        
        // Clean up the input
        input = input.trim();
        
        // Try parsing with Day.js
        let parsed;
        try {
            parsed = this._parseWithDayjs(input);
        } catch (err) {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                `unable to parse date/time: expected formats like '2006-01-02T15:04:05Z' or '01/02/2006 3:04 PM', got '${input}'`,
                input,
                err
            );
        }
        
        // Check if input explicitly specified timezone
        const hasTimezone = input.includes('Z') || input.includes('+') || 
                           (input.includes('T') && input.lastIndexOf('-') > input.lastIndexOf('T')) ||
                           input.toUpperCase().endsWith('UTC');
        
        if (!hasTimezone) {
            // Apply default timezone - interpret the parsed time as being in the default timezone
            try {
                // Use Day.js to create a time in the default timezone
                parsed = dayjs.tz(parsed.format('YYYY-MM-DD HH:mm:ss'), this.config.defaultTimezone);
            } catch (err) {
                throw newDateTimeError(
                    ERROR_TYPES.INVALID_TIMEZONE,
                    `invalid default timezone: ${this.config.defaultTimezone}`,
                    input,
                    err
                );
            }
        }
        
        // Convert to native Date object and store timezone info
        const result = parsed.toDate();
        
        // Store original timezone info for formatting
        if (input.includes('Z')) {
            result._originalTimezone = 'UTC';
        } else if (hasTimezone && input.includes('T') && input.lastIndexOf('-') > input.lastIndexOf('T')) {
            // For timezone offsets like -05:00, we need to map them to timezone names
            // This matches Go's behavior which shows "EST" instead of "-05:00"
            const tzPart = input.substring(input.lastIndexOf('-'));
            result._originalTimezone = this._mapOffsetToTimezone(tzPart);
        } else if (hasTimezone && input.includes('+')) {
            // For positive timezone offsets like +08:00
            const tzPart = input.substring(input.lastIndexOf('+'));
            result._originalTimezone = this._mapOffsetToTimezone('+' + tzPart);
        } else if (!hasTimezone) {
            result._originalTimezone = this.config.defaultTimezone;
        }
        
        return result;
    }
    
    /**
     * Parses a date-only string and returns a Date at midnight in the default timezone
     * @param {string} input - The date string to parse
     * @returns {Date} - Parsed Date object at midnight
     * @throws {DateTimeError} - If parsing fails
     */
    parseDate(input) {
        if (!input || typeof input !== 'string') {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                'empty date input',
                input,
                null
            );
        }
        
        input = input.trim();
        
        // Try parsing with Day.js using various date formats
        let parsed;
        const dateFormats = [
            'YYYY-MM-DD',
            'MM/DD/YYYY',
            'M/D/YYYY',
            'MMMM D, YYYY',
            'MMM D, YYYY',
            'D MMMM YYYY',
            'D MMM YYYY'
        ];
        
        try {
            // Try each format
            for (const format of dateFormats) {
                parsed = dayjs(input, format, true); // strict parsing
                if (parsed.isValid()) {
                    break;
                }
            }
            
            if (!parsed || !parsed.isValid()) {
                throw new Error('No valid format found');
            }
        } catch (err) {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                `unable to parse date: expected formats like '2006-01-02' or '01/02/2006', got '${input}'`,
                input,
                err
            );
        }
        
        // Set to midnight in the default timezone
        try {
            // Create a date at midnight in the default timezone using Day.js
            const result = dayjs.tz(parsed.format('YYYY-MM-DD') + ' 00:00:00', this.config.defaultTimezone).toDate();
            result._originalTimezone = this.config.defaultTimezone;
            return result;
        } catch (err) {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_TIMEZONE,
                `invalid default timezone: ${this.config.defaultTimezone}`,
                input,
                err
            );
        }
    }
    
    /**
     * Parses a time-only string and combines it with the given date
     * @param {string} input - The time string to parse
     * @param {Date} date - The date to combine with
     * @returns {Date} - Combined Date object
     * @throws {DateTimeError} - If parsing fails
     */
    parseTime(input, date) {
        if (!input || typeof input !== 'string') {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                'empty time input',
                input,
                null
            );
        }
        
        if (!date || !(date instanceof Date)) {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                'invalid date parameter',
                input,
                null
            );
        }
        
        input = input.trim();
        
        // Time-only formats for Day.js
        const timeFormats = [
            'HH:mm:ss',
            'HH:mm',
            'h:mm:ss A',
            'h:mm A',
            'h:mm:ssA',
            'h:mmA'
        ];
        
        let parsed;
        try {
            // Try each time format
            for (const format of timeFormats) {
                parsed = dayjs(input, format, true); // strict parsing
                if (parsed.isValid()) {
                    break;
                }
            }
            
            if (!parsed || !parsed.isValid()) {
                throw new Error('No valid format found');
            }
        } catch (err) {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                `unable to parse time: expected formats like '15:04' or '3:04 PM', got '${input}'`,
                input,
                err
            );
        }
        
        // Combine with the provided date, preserving the date's timezone context
        const baseDayjs = dayjs(date);
        const combined = baseDayjs
            .hour(parsed.hour())
            .minute(parsed.minute())
            .second(parsed.second())
            .millisecond(parsed.millisecond());
        
        const result = combined.toDate();
        result._originalTimezone = date._originalTimezone;
        return result;
    }
    
    /**
     * Parse input using Day.js with multiple format attempts
     * @param {string} input - The input string to parse
     * @returns {dayjs.Dayjs} - Parsed Day.js object
     * @throws {Error} - If all formats fail
     */
    _parseWithDayjs(input) {
        // First try standard Day.js parsing (handles ISO formats automatically)
        let parsed = dayjs(input);
        if (parsed.isValid()) {
            return parsed;
        }
        
        // Try specific formats
        const formats = [
            // ISO8601/RFC3339 formats
            'YYYY-MM-DDTHH:mm:ssZ',
            'YYYY-MM-DDTHH:mm:ss',
            'YYYY-MM-DDTHH:mm',
            
            // Date only formats
            'YYYY-MM-DD',
            'MM/DD/YYYY',
            'M/D/YYYY',
            'MMMM D, YYYY',
            'MMM D, YYYY',
            'D MMMM YYYY',
            'D MMM YYYY',
            
            // Combined date/time formats
            'YYYY-MM-DD HH:mm:ss',
            'YYYY-MM-DD HH:mm',
            'MM/DD/YYYY HH:mm:ss',
            'MM/DD/YYYY HH:mm',
            'MM/DD/YYYY h:mm A',
            'MMMM D, YYYY h:mm A',
            'MMMM D, YYYY [at] h:mm A'
        ];
        
        for (const format of formats) {
            parsed = dayjs(input, format, true); // strict parsing
            if (parsed.isValid()) {
                return parsed;
            }
        }
        
        throw new Error('No valid format found');
    }
    
    /**
     * Parses a date/time string and applies the specified timezone
     * @param {string} input - The date/time string to parse
     * @param {string} timezone - The timezone to apply
     * @returns {Date} - Parsed Date object in specified timezone
     * @throws {DateTimeError} - If parsing fails
     */
    parseDateTimeWithTimezone(input, timezone) {
        // If no timezone specified, use normal parsing
        if (!timezone) {
            return this.parseDateTime(input);
        }
        
        try {
            // Validate timezone first using Intl API (more reliable than Day.js)
            try {
                new Intl.DateTimeFormat('en-US', { timeZone: timezone });
            } catch (err) {
                throw newDateTimeError(
                    ERROR_TYPES.INVALID_TIMEZONE,
                    `invalid timezone: ${timezone}: unknown time zone ${timezone}`,
                    input,
                    err
                );
            }
            
            // Check if input explicitly specified timezone
            const cleaned = input.trim();
            const hasTimezone = cleaned.includes('Z') || cleaned.includes('+') || 
                               (cleaned.includes('T') && cleaned.lastIndexOf('-') > cleaned.lastIndexOf('T')) ||
                               cleaned.toUpperCase().endsWith('UTC');
            
            if (hasTimezone) {
                // Input already has timezone, parse normally then convert to target timezone
                const parsed = dayjs(cleaned);
                if (!parsed.isValid()) {
                    throw newDateTimeError(
                        ERROR_TYPES.INVALID_FORMAT,
                        `unable to parse date/time: expected formats like '2006-01-02T15:04:05Z' or '01/02/2006 3:04 PM', got '${cleaned}'`,
                        cleaned,
                        null
                    );
                }
                
                const converted = parsed.tz(timezone);
                const result = converted.toDate();
                result._originalTimezone = timezone;
                return result;
            } else {
                // Input has no timezone, interpret it as being in the specified timezone
                // This is the key: we want to treat the wall clock time as being in the target timezone
                
                let parsed;
                try {
                    parsed = this._parseWithDayjs(cleaned);
                } catch (err) {
                    throw newDateTimeError(
                        ERROR_TYPES.INVALID_FORMAT,
                        `unable to parse date/time: expected formats like '2006-01-02T15:04:05Z' or '01/02/2006 3:04 PM', got '${cleaned}'`,
                        cleaned,
                        err
                    );
                }
                
                // Use Day.js to interpret the wall clock time as being in the target timezone
                // This matches Go's time.Date() behavior exactly
                const result = dayjs.tz(parsed.format('YYYY-MM-DD HH:mm:ss'), timezone).toDate();
                result._originalTimezone = timezone;
                return result;
            }
        } catch (err) {
            if (err instanceof DateTimeError) {
                throw err;
            }
            throw newDateTimeError(
                ERROR_TYPES.INVALID_TIMEZONE,
                `invalid timezone: ${timezone}: ${err.message}`,
                input,
                err
            );
        }
    }
    
    /**
     * Parses separate date and time fields (for backward compatibility)
     * @param {string} date - The date string
     * @param {string} timeStr - The time string
     * @param {string} timezone - The timezone string
     * @returns {Date} - Combined Date object
     * @throws {DateTimeError} - If parsing fails
     */
    parseLegacyDateTimeFields(date, timeStr, timezone) {
        if (!date) {
            throw newDateTimeError(
                ERROR_TYPES.INVALID_FORMAT,
                'date field is required',
                date,
                null
            );
        }
        
        // Parse the date
        let parsedDate = this.parseDate(date);
        
        // If time is provided, parse and combine it
        if (timeStr) {
            parsedDate = this.parseTime(timeStr, parsedDate);
        }
        
        // Apply timezone if specified
        if (timezone) {
            try {
                // Interpret the parsed date as being in the specified timezone using Day.js
                const dayjsDate = dayjs(parsedDate);
                const result = dayjs.tz(dayjsDate.format('YYYY-MM-DD HH:mm:ss'), timezone).toDate();
                result._originalTimezone = timezone;
                parsedDate = result;
            } catch (err) {
                throw newDateTimeError(
                    ERROR_TYPES.INVALID_TIMEZONE,
                    `invalid timezone: ${timezone}`,
                    timezone,
                    err
                );
            }
        }
        
        return parsedDate;
    }
    
    /**
     * Map timezone offset to timezone name (to match Go's behavior)
     * @private
     * @param {string} offset - Timezone offset like "-05:00" or "+08:00"
     * @returns {string} - Timezone name or the original offset
     */
    _mapOffsetToTimezone(offset) {
        // Common timezone offset mappings (for January - EST/PST are in standard time)
        const offsetMap = {
            '-05:00': 'America/New_York',  // EST
            '-08:00': 'America/Los_Angeles', // PST
            '-06:00': 'America/Chicago',   // CST
            '-07:00': 'America/Denver',    // MST
            '+00:00': 'UTC',
            'Z': 'UTC'
        };
        
        return offsetMap[offset] || offset;
    }

}